spec/p2p/p2p.qnt

```diff
@@ -0,0 +1,140 @@
+    var Environment: Set[Address]
+ 
+
+    pure val MaxNumOutboundPeers = 5
+    pure val MaxNumInboundPeers = 10
+
+    def Peers = peerState.keys()
+
+    def peersOutbound = size(Peers.filter(a => peerState.get(a) == "outbound"))
+    def peersInbound = size(Peers.filter(a => peerState.get(a) == "inbound"))
+    def peersDialing = size(Peers.filter(a => peerState.get(a) == "dialing"))
+
+    def require(pred:bool):bool = pred
+    //pure def putset (m,s)
+
+    // triggered by "the network"
+    action loseConnection(a:Address):bool = all{
+        require(connectionState.get(a) != "closed"),
+        connectionState' = connectionState.set(a, "closed"),
+        all { //unchanged
+            AddressBook' = AddressBook,
+            peerState' = peerState,
+            Environment' = Environment,
+            IsUnconditional' = IsUnconditional,
+        }
+    }
+
+    action transportAccept(a: Address):bool = all {
+        or {
+            peersInbound < MaxNumInboundPeers,
+            IsUnconditional.contains(a)
+        },
+        peerState' = peerState.put(a, "inbound"),  //shall we check if it is already outbound and open?
+        connectionState' = connectionState.put(a, "open"),
+        AddressBook' = AddressBook.union(Set(a)) ,
+        all { //unchanged
+            Environment' = Environment,
+            IsUnconditional' = IsUnconditional,
+        }
+    }
+
+    // triggered by a reactor
+    action stopPeerForError(a:Address):bool = all {
+        peerState' = Peers.exclude(Set(a)).mapBy(x => peerState.get(x)),
+        connectionState' = Peers.exclude(Set(a)).mapBy(x => connectionState.get(x)),
+        all { //unchanged
+            AddressBook' = AddressBook,
+            Environment' = Environment,
+            IsUnconditional' = IsUnconditional,
+        }
+    }
+        // inform all reactors then
+    
+
+    action ensurePeers(a:Address):bool =
+        all {
+                require(peersOutbound + peersDialing < MaxNumOutboundPeers),
+                require(not(Peers.contains(a))),
+                all {
+                    // I guess we don't remove AddressBook' = AddressBook.exclude(Set(a)),
+                    //dialPeer(addr)
+                    peerState' = peerState.put(a, "dialing"), 
+                    connectionState' = connectionState.put(a, "init"),
+                    all { //unchanged
+                        AddressBook' = AddressBook,
+                        Environment' = Environment,
+                        IsUnconditional' = IsUnconditional,
+                    }
+                }
+            }
+
+    action dialPeers(S:Set[Address]):bool = // check what the code actually is doing. Now we don't override existing peers in this spec
+        all {
+            peerState' = Peers.union(S).mapBy(a => if (Peers.contains(a)) peerState.get(a) else "dialing"), 
+            connectionState' = Peers.union(S).mapBy(a => if (Peers.contains(a)) connectionState.get(a) else "init"),
```

Comment:
We should have `dialPeer` also here. I am assuming that this method either set a peer to the "outbound" state or produce an error, right?

Reply:
Since this method is invoked from "outside", the addresses are also added to the Address Book.